File: src/services/LocationService.ts
import axios from 'axios';

export interface Location {
    id: string;
    name: string;
    zipcode: string;
    canton: string;
}

export class LocationService {
    private static API_URL = 'https://api.post.ch/ZIP';
    private static cache = new Map<string, Location[]>();

    static async searchLocations(query: string): Promise<Location[]> {
        const cleanQuery = query.trim();
        
        if (this.cache.has(cleanQuery)) {
            return this.cache.get(cleanQuery)!;
        }

        try {
            const isNPA = /^\d+$/.test(cleanQuery);
            
            const response = await axios.get('https://api.swisspost.ch/locations/v1/localities/search', {
                params: {
                    q: cleanQuery,
                    limit: 10,
                    type: isNPA ? 'zip' : 'locality',
                    language: 'fr'
                },
                headers: {
                    'Accept': 'application/json',
                    'Accept-Language': 'fr'
                }
            });

            if (!response.data?.localities) {
                return [];
            }

            const locations = response.data.localities
                .filter((loc: any) => loc.zip && loc.name)
                .map((loc: any) => ({
                    id: `${loc.zip}-${loc.name}`,
                    name: loc.name,
                    zipcode: loc.zip,
                    canton: loc.canton
                }));

            this.cache.set(cleanQuery, locations);
            return locations;
        } catch (error) {
            console.error('Erreur de recherche:', error);
            return [];
        }
    }
}

File: data/src/bot/services/location_service.py
import requests
from dataclasses import dataclass

@dataclass
class Location:
    name: str
    zipcode: str
    canton: str

class LocationService:
    @staticmethod
    def search_locations(query: str) -> list[Location]:
        try:
            # Utilisation de l'API officielle des communes suisses
            response = requests.get(
                'https://api.ch.ch/addresses/locations',
                params={
                    'searchtext': query,
                    'lang': 'fr',
                    'limit': 10
                },
                headers={
                    'Accept': 'application/json'
                }
            )
            
            if response.status_code != 200:
                return []

            results = response.json()
            return [
                Location(
                    name=loc['name'],
                    zipcode=loc['zip'],
                    canton=loc['canton']
                )
                for loc in results
            ]
        except Exception as e:
            print(f"Erreur de recherche: {e}")
            return []
